core/lib/web3_decl/src/error.rs

```diff
@@ -42,3 +43,65 @@ pub enum Web3Error {
     #[error("Tree API is not available")]
     TreeApiUnavailable,
 }
+
+use std::{collections::HashMap, error, fmt, fmt::Debug};
+
+#[derive(Debug)]
+pub struct RpcErrorWithDetails {
+    pub inner_error: RpcError,
+    pub method: String,
+    pub args: HashMap<String, String>,
+}
+impl error::Error for RpcErrorWithDetails {}
+
+impl fmt::Display for RpcErrorWithDetails {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let args_formatted: Vec<String> =
+            self.args.iter().map(|(k, v)| format!("{k}: {v}")).collect();
+        write!(
+            formatter,
+            "JsonRPC request: {} with args: {} failed because of: {}",
+            self.method,
+            args_formatted.join(", "),
+            self.inner_error
+        )
+    }
+}
+
+pub trait EnrichRpcError<T> {
+    fn rpc_context(self, method: &str) -> Result<T, RpcErrorWithDetails>;
+}
+
+impl<T> EnrichRpcError<T> for Result<T, RpcError> {
+    fn rpc_context(self, method: &str) -> Result<T, RpcErrorWithDetails> {
+        match self {
+            Ok(t) => Ok(t),
+            Err(error) => Err(RpcErrorWithDetails {
+                inner_error: error,
+                method: method.to_string(),
+                args: HashMap::default(),
+            }),
+        }
+    }
+}
+
+pub trait WithArgRpcError<T> {
```

Comment:
Bikeshedding: 2 distinct traits is why I haven't proposed this option when discussing how best to implement this functionality; IMO, it looks mildly awkward. If you really think that it's the best option, I won't oppose (personally, I'd like the `Future`-wrapping approach better; it has a single trait and also allows tracking long-executing requests).

Reply:
I decided to just leave it as it is